src/lib.rs

```diff
@@ -160,8 +161,86 @@ use bitcoin_hashes::Hash;
 
 /// An ECDSA signature
 #[derive(Copy, Clone, PartialEq, Eq)]
-pub struct Signature(ffi::Signature);
+#[cfg_attr(feature="json", derive(schemars::JsonSchema))]
+pub struct Signature(
+    #[cfg_attr(feature="json", schemars(schema_with="schemas::signature"))]
+    ffi::Signature);
+
+/// Custom JSON Schemas for interior types
+#[cfg(feature="json")]
+pub mod schemas {
+    use schemars::schema::{Schema, SchemaObject};
+    use schemars::{gen::SchemaGenerator, JsonSchema};
+    const HEX_REGEX : &'static str = "[0-9a-fA-F]+";
+
+    /// schema for Signature
+    /// Max length 72*2, no min length. Must be hex, upper or lower.
```

Comment:
upper or lower: Non-canonical serializations have created a number of crypto issues in the past. Sure, users not supposed to use the hex serialization for anything than deserialize it but can we make this either upper or lower? (Same for the other types)

no min length:
The regex enforces a min length of half a byte, see above.

max length:
I'm not entirely sure about this. Invalid signatures can be longer.

Reply:
The reason to make it upper or lower is because it could be coming from another source that chose lower when we chose upper.

Good catch on the min length, will accept the other patch. I think there is a min length we could support, it's something like 5 or 6 bytes?

w.r.t. decoding an invalid signature length, that's an interesting point. I'd err on the side of making that against the schema, as we want it to validate as much as possible, but perhaps there is a use for sending an invalid signature? I'd think that should be another explicit type though...

Reply:
> The reason to make it upper or lower is because it could be coming from another source that chose lower when we chose upper.

Ok sure. My suggestion is to err on the side of caution and prefer non-malleability over convenience here.

> 
> Good catch on the min length, will accept the other patch. I think there is a min length we could support, it's something like 5 or 6 bytes?

Hm, maybe, but why not leave at 0 bytes?

> 
> w.r.t. decoding an invalid signature length, that's an interesting point. I'd err on the side of making that against the schema, as we want it to validate as much as possible, but perhaps there is a use for sending an invalid signature? I'd think that should be another explicit type though...

Well I think restricting it to 72 bytes is fine. (There may be a use for sending invalid signatures but you could just fail to parse the signature before you can even put it in a JSON. And if you really want to send it then, you can send a canonically invalid signature instead, e.g., 72 zeros, or the empty string.) We just need to check that this does not create any issues with the rest of library, which might not assume 72 bytes maximum I think the library assumes this, I just haven't checked right now.

Reply:
| Ok sure. My suggestion is to err on the side of caution and prefer non-malleability over convenience here.

My opinion is that we should match the rules of the from_hex function, which allows for mixed upper or lowercase of no more than expected length. To not match this would be surprising.

| Hm, maybe, but why not leave at 0 bytes?

It could help differentiate an optional field being left empty v.s. a canonical null signature. I think there's some minimum null signature in DER, which is something like 6 bytes.

Reply:
> | Ok sure. My suggestion is to err on the side of caution and prefer non-malleability over convenience here.
> 
> My opinion is that we should match the rules of the from_hex function, which allows for mixed upper or lowercase of no more than expected length. To not match this would be surprising.

Hm yes. My concern is that someone will get the idea that this encoding is unique and put this serialization into a hash function somewhere to derive some values, and then you can run into all kind of trouble.

> | Hm, maybe, but why not leave at 0 bytes?
> 
> It could help differentiate an optional field being left empty v.s. a canonical null signature. I think there's some minimum null signature in DER, which is something like 6 bytes.

So what is the field supposed to contain?  Valid DER encodings of two integers? Arbitrary byte strings that may or may not be valid DER? Valid DER encodings of any data structure?

Reply:
There are endless ways someone can fuck anything up; e.g., they could assume that it's big endian or little endian when it's not. The only proper thing to do is to decode the hex into a byte buffer before hashing. Otherwise if you want to add a restriction here, I think it's unacceptable that our hex de-serialization supports mixed case, we should require all upper or lower case. JSONSchema is *only* for basic structural checking and validation. This is quite useful in many contexts! But it does not embed full validation logic. It's a mistake to use it as such. JSONSchema doesn't guarantee any consistency! E.g., {'a':1, 'b': 2} and {'b':2, 'a':1} are both valid if one is valid under (I think) all schemas, there's no ordering of fields in a json. Hashing is not an appropriate use case.


| So what is the field supposed to contain? Valid DER encodings of two integers? Arbitrary byte strings that may or may not be valid DER? Valid DER encodings of any data structure?

Looking at the from_hex code, which I guess should be the canon here, along with from_str and from_der, an empty string will return InvalidSignature, which our goal with JsonSchema is eliminate as many errors as possible, so empty can be disallowed as it will unconditionally not decode.

Looking through the source, we eventually are at https://github.com/bitcoin-core/secp256k1/blob/39198a03eaa33d5902b16d3aefa7c441232f60fb/src/ecdsa_impl.h#L153 which looks like a minimum decodable signature is 6 bytes if I've counted correctly?

That's all we really care about, that we can validly decode the string into a Signature -- validation of the Signature itself is left to the user.

Reply:
> There are endless ways someone can fuck anything up; e.g., they could assume that it's big endian or little endian when it's not. 

Hm but that's an entirely different way of fucking up, namely in parsing. If you implement parsing, then you of course need to do it correctly. But then you're implementing a low-level ECC operations. Of course you can fuck up there in million ways. But hashing (or doing other operations) on a serialization does not require understanding the values, it's more likely to happen in user code.

> JSONSchema is _only_ for basic structural checking and validation. This is quite useful in many contexts! But it does not embed full validation logic.

Sure, but why then require a minimum of 6 bytes? 

I spent some thoughts on this. Currently a byte string can fail two be a valid signature in two ways. First, it can fail to parse. Second, it can fail verification (with respect to some given pubkey and message).  This is already one more way more than really necessary, but that's how it is for ECDSA. (We improved this in BIP340 for Schnorr signatures: There's simply no parsing step, so you need to handle fewer errors as a user and check fewer return values.) Now if we require a minimum of 6 bytes, we add a third way in which byte strings can be invalid ECDSA signatures, namely if they're too short. This requires even more error handling. I think that's a step in the wrong direction. I'm ok with an upper bound of 72 bytes because this is commonly implemented (whereas a check for 6 bytes is implemented nowhere else -- at least not that I'm aware of) and we need an upper bound somewhere to protect against DoS. Funnily enough, not even the C code has restrictions here. You can pass a 4GB string to `secp256k1_ecdsa_sig_parse` and it will try to parse it.

Reply:
I just think it's far-fetched; we should never be hashing a json and expecting non malleability full stop. You're also asking for the rules to be divergent from what will still successfully deserialize, which is all that JSONSchema is designed to help you with.



w.r.t the second point, we *do* currently fail to parse a signature less than 6 bytes long, so it is an existing error. JSONSchema isn't making new errors, it's just preventing you from passing a value that would already return a parse error. Not a new error, existing error. If that error didn't already exist I would agree with you.

Reply:
btw it would even be possible to make sure the signature bears the proper 0x30 prefix in the regex for signature too, which might be nice?

Reply:
> I just think it's far-fetched; we should never be hashing a json and expecting non malleability full stop.

Sure, hashing the JSON is stupid. What I had in mind is hashing a hex-encoded signature. There's nothing wrong with that if the encoding is unique.

> w.r.t the second point, we _do_ currently fail to parse a signature less than 6 bytes long, so it is an existing error. JSONSchema isn't making new errors, it's just preventing you from passing a value that would already return a parse error. Not a new error, existing error. If that error didn't already exist I would agree with you.

Ok, I see. So you said it's about interoperability with other programs. 

Consider a scenario where some remote machine A receives a signature from the network and forwards it B, running rust-secp. Now A  (which may be using an entirely different piece of software written in another language) needs to deal with restrictions such as the signature being at least 6 bytes long and/or the signature starting with 0x30. Why should A bother checking these conditions if the signature anyway can fail later when B parses it again?

Reply:
> Sure, hashing the JSON is stupid. What I had in mind is hashing a hex-encoded signature. There's nothing wrong with that if the encoding is unique.

This is fine, just make sure that you re-encode the signature according to your local rules. Don't blindly encode user input without making sure it is canonical. 

> Why should A bother checking these conditions if the signature anyway can fail later when B parses it again?

Here's a library I use that uses JSONSchema. https://github.com/rjsf-team/react-jsonschema-form

It can automatically build web forms from JSONSchema and do client side validation which helps users correctly input data for a given type, and inform users of simple errors (e.g. -- pattern doesn't match or value not long enough) without having to submit it to the server.